README

  MudPRO: An advanced client for the online game MajorMUD
  Copyright (C) 2002, 2003, 2004  David Slusky

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


 CREATING A NEW PROFILE
========================

Once you have unpacked the archive, you will need to setup a profile for
your character to use. To do this, first make a copy of the default profile:

  cp -a profile/default profile/khamsin

Then edit the character.conf file to the desired settings. Be sure that the
profile path is set correctly. If you are running the program from the
directory where you extracted it, then the profile path for khamsin would
be "profile/khamsin". Don't forget, *nix is case sensitive!


 RUNNING THE APPLICATION
=========================

When launching MudPRO you must specify the profile to use, such as:

  ./mudpro -p profile/khamsin/character.conf

The preceeding dot-slash indicates that the program resides in the current
directory. Many distros do not normally include the current directory within
a users command search path (for security reasons). If you receive the message
"command not found", use the dot-slash or change your PATH environment variable.


 SOME THINGS YOU NEED TO KNOW
==============================

* CONFIGURING THE CLIENT DATA

To make the most of MudPRO, you will need to edit the various data files in
your character profile using a text editor. Each of these files contains a
brief comment section explaining how to configure it. Not all of the options
avilable are covered however. For now, let me know if there is something not
listed you'd like to customize. You can also see more of the options available
by viewing the SAMPLE profile.

Most of the files use a format in which leading whitespace is significant,
similar to the Python programming language. Individual records are denoted
by a declaration placed at the beginning of a new line with no whitespace
before it. Optional parameters are then placed on subsequent lines and must
be indented using (any number of) spaces or tabs. One or more blank lines
must come between declarations to separate them. Lines beginning with a '#'
are considered comments and will be ignored.

#
# Here's an example of what two definitions might look like
# in the items.db control list:
#

Item = "darkwood ring"
  SetFlag = AutoDrop

Item = "oaken staff"
  SetFlag = AutoGet
  Limit   = 10

TIP: You can make changes to any of the data files while the program is
already running. Whenever you do, the file will automatically be reloaded
for you. That is, unless you have disabled conf polling :)


* UPDATING TO A NEWER VERSION

When updating to a newer version, be sure to close down any sessions that
are active. Then unpack the active to the same location and overwrite the
old files. Keep in mind that this only affects the default profile. If you
want to update your character(s), you can use the update-profile script:

  ./update-profile profile/khamsin

NOTE: using the update-profile script will overwrite any changes you've made
to the monster.db and parse data files. Use it with caution!


 SOME ISSUES YOU MIGHT ENCOUNTER
=================================

* ANSI EMULATION PROBLEMS

If certain elements inside the terminal area or in the UI look incorrect, it's
because the font you're using doesn't contain the IBM extended character set.
Unfortunately, this will probably be the case as most fonts don't. If you are
unable to locate a suitable font (I will try to help), there are a couple of
workarounds. First, try launching the program like so:

  TERM=linux ./mudpro -p profile/khamsin/character.conf

This might at least correct the line drawing characters used in the UI.
Failing that, you could enable the ASCII-only line style by launching the
program like so:

  ./mudpro -p profile/khamsin/character.conf --line-style 3

Neither of these workarounds will completely solve the problem however.


* TERMINAL EMULATOR ISSUES

Many terminal emulators on Linux use a rather bright and unappealing color
palette by default. This could make things such as the menu items difficult to
see properly. Unfortunately, there is nothing I can do to resolve the problem
from within my program. You will need to adjust the colors using your console.

If you are unable to access the drop-down menus (such as the ALT-A automap
menu), try using ESC-A instead. If this works, you will need to reconfigure
how your terminal handles the meta (ALT) key. Xterm in particular seems to
exhibit this problem, I would recommend using something better.


* HP/MA TRACKING IS INCORRECT

At the present time, MudPRO only understands the default status prompt. If
you are using a custom prompt HP/MA updates will not be read corretly. Adding
support for custom prompts is somthing I would like to add in the future, but
for now please use the default one (eg: set statline full)


* UNHANDLED MESSAGES

If you are seeing unusual or undesired behavior, the most likely cause is an
unhandled game message. Take a look at the messages you have received in the
past few rounds. See anything that would affect the situation? Check to see
if those messages are handled correctly (or at all!) in the parse data files.

When you do find a message that needs to be handled differently, be sure to
let me know about it. That way the changes can be included in the future.

Additionally, you could create a Report Log to see if that helps shed some
light on any problems you might be having. To do this, press ALT-H (for the
help menu) followed by R.


* RESIZING CRASHES

If you resize your terminal window too small (and by that I mean really small!)
it could hang the application or lead to a crash. I am working to resolving this
issue.


 AUTOMAPPING
=============

Congratulations on making it this far ;) While I've tried to make using the
automapper as easy and intuitive as possible, there are still some gotchas.
If you need to make any additions to the db files as discussed below, be sure
to pass along your changes so they can be included with future releases.


* ADJUSTING COORDINATES

Sometimes it is necessary to manually adjust the XYZ coordinates for a given
location. Hit ALT-A (for the automapper menu) and then C to begin making an
adjustment. In the Mapview OSD title area you will see a prompt asking you to
select an axis.

To select the X axis, press X. Use the + and - keys to make adjustments.
You can select another axis at any time by pressing X, Y, or Z. When you are
finished, press ALT-A and C again to complete the adjustment.

One example where you will need to make adjustments is the Grand Stair. It
is necessary to lower the Z axis as you descend, otherwise the rooms will
overlap incorrectly.


* CUSTOM SECRET EXITS

Most secret exits are easily identified because the exit is explicitly named
when it's discovered (eg: You found an exit east!). Many secrets on the other
hand use a custom message without any indication as to what direction the exit
was in. Worse yet, some of these messages are reused in different places!

Because of this situation, the automapper needs a little help in learning
which exit has been uncovered. The secrets.db file in the parse subdirectory
provides the needed information. You should take a moment to look it over,
don't worry it's nothing complicated.

If you encounter a custom secret while automapping, take a look at the
Mapview OSD. If you do not see the exit displayed there, you will _NEED_ to
add an entry for it. Without this information the automapper cannot join the
locations correctly. Once you have added an appropriate entry, reactivate
the secret exit as before and it should then be recognized.


* SECRET PATHS

Just like custom secrets, the secret paths that transport you places such
as "go hole" leave out some information the automapper really needs. There
are two different scenarios you will need to deal with, but more importantly
you must notify the automapper when a special movement has taken place.

In the movement.db file there is a group of messages along these lines:

"You push through the dense foliage, and walk onto a small path."
"You clamber into the burnt stump."
"You climb into the pitch-dark hole."
    Automap:        SpecialMovement

As you can see, each of these messages invokes the Automap:SpecialMovement
action. If you become lost after following a path, then you will need to
make sure the corresponding message has been added to that group properly.

Now, as I mentioned, there are two kinds of secret path scenarios. The first
kind also has a secret path leading in the opposite direction. In the Darkwood
Forest near the Blue Tower there are a couple of these. You need to use
"go path" in both directions. In this scenario, your work is done. There is

ChangeLog
Release 0.91r (?)

 - Added "Smash" as a WeaponAttack modifier (overrides Bash).
 - Visiting multiple locations to open an exit is now possible.
 - Lots of misc party code fixes and additions
 - Alternate telnet port number can now be specified.
 - The character profile command-line option has been changed from -p to -f.

Release 0.91q (05-02-2004)

 - Added PartyWait and ParInterval character options.
 - Now responds to @wait as leader
 - Internal tracking of followers implemented
 - Some autoroam double-back corrections
 - Added 'Reserve' item option, this allows you to define an amount of an item
   to hold in reserve before stashing. Default is 0.
 - Automap position no longer reset on cleanup events. It's pretty safe to
   assume you will appear in the same location in this context. So now there
   is little worry that you will accidently wander someplace you shouldn't.

Release 0.91p (03-12-2004)

 - Added @stop/@rego remote commands.
 - Added set of @auto-toggle remotes to control player toggles
 - Implemented system to request assistance from party members.
 - Added "Prefix" option to use when talk is set to slow.
 - Added another ascii line style (#4).

Release 0.91o (03-10-2004)

 - Added the ability to capture the current session to file. You can enable it
   through the toggle menu or by using the -r command-line option.
 - Combat strategy is now recalculated as necessary when multiple monsters
   enter the room at once.

Release 0.91n (03-07-2004)

 - Corrected combat regression so that you will now break properly after
   using room spells again.
 - Added ModHealth and ModStats item flags to signal changes in player
   attributes when (dis)arming items. Setting ModHealth for example will
   cause MudPRO to recheck your health.
 - Fixed potential problem when sys goto is used while running.

Release 0.91m (03-04-2004)

 - Rooms with long names are now shortened in guidebook.
 - Default automap.db now included, thanks Akc99!
 - Fixed problem parsing "You notice {player} sneak in from the {direction}."
 - EXIT_FLAG_ROOMCLEAR implementation should now be complete.

Release 0.91l (02-25-2004)

 - Corrected menubar problem introduced in the previous version.

Release 0.91k (02-25-2004)

 - Added EXIT_FLAG_ROOMCLEAR exit flag to force clearing the room before
   crossing a particular exit (overrides disabled auto-attack).
 - Improved target override implementation.
 - Numerous code fixes to facilitate Win32 port (Thanks Locke).
 - Added support for "Two handed" weapons in inventory.

Release 0.91j (02-23-2004)

 - Made transition to glib 2.x from 1.2.x.
 - Now using glib for directory reading, instead of opendir/readdir.

Release 0.91i (02-23-2004)

 - Temporarily removed code allowing uppercase input to access menus,
   it caused conflicts in some cases.

Release 0.91h (02-23-2004)

 - Switched to using g_get_current_time() instead of gettimeofday()
 - Added FollowAttack option to attack the same enemies as leader
 - Added MonsterCheck option to force checking room before engaging
 - Drop-down menus now accessble while capslock enabled
 - Warn user when attempting to "use {key}" that isn't in inventory

Release 0.91g (02-20-2004)

 - Corrected pattern that extracts data from "who" list.
 - Initial party support added, including several @commands.
 - Added parse/custom.db for storing user-defined patterns. This file will
   not be overwritten by update-profile.
 - Adopted new ChangeLog format that should be easier to read through.

Release 0.91f (02-18-2004)

 - Disabled access to autoroaming while automapping and vice-versa.
 - Fixed character command list not being reset properly.

Release 0.91e (02-17-2004)

 - Fixed potential double free errors in item.c and player.c

Release 0.91d (02-16-2004)

 - Initial support for other players added.

Release 0.91c (02-15-2004)

 - Added notice in README regarding status prompts.
 - GuideBook items are now removed when the automap database is reset.
 - Monster target list should no longer accumulate incorrectly in certain.
   circumstances.

Release 0.91b (02-14-2004)

 - Fixed potential crash when creating report log.
 - Fixed problem when connecting by IP address.

Release 0.91a (02-13-2004)

 - Initial public release.


src/item.c
/*  MudPRO: An advanced client for the online game MajorMUD
 *  Copyright (C) 2002, 2003, 2004  David Slusky
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <ctype.h>
#include <stdlib.h>
#include <string.h>

#include "automap.h"
#include "character.h"
#include "combat.h"
#include "command.h"
#include "item.h"
#include "navigation.h"
#include "mudpro.h"
#include "timers.h"
#include "terminal.h"
#include "utils.h"

key_value_t item_flags[]=
{
	{ "Usable",    ITEM_FLAG_USABLE },
	{ "Currency",  ITEM_FLAG_CURRENCY },
	{ "Light",     ITEM_FLAG_LIGHT },
	{ "Stash",     ITEM_FLAG_STASH },
	{ "AutoGet",   ITEM_FLAG_AUTO_GET },
	{ "AutoDrop",  ITEM_FLAG_AUTO_DROP },
	{ "AutoArm",   ITEM_FLAG_AUTO_ARM },
	{ "ModHealth", ITEM_FLAG_MOD_HEALTH },
	{ "ModStats",  ITEM_FLAG_MOD_STATS },
	{ NULL, 0 }
};

key_value_t equip_flags[]=
{
	{ "Arms",        EQUIP_FLAG_ARMS },
	{ "Back",        EQUIP_FLAG_BACK },
	{ "Ears",        EQUIP_FLAG_EARS },
	{ "Finger",      EQUIP_FLAG_FINGER },
	{ "Feet",        EQUIP_FLAG_FEET },
	{ "Hands",       EQUIP_FLAG_HANDS },
	{ "Head",        EQUIP_FLAG_HEAD },
	{ "Legs",        EQUIP_FLAG_LEGS },
	{ "Neck",        EQUIP_FLAG_NECK },
	{ "Readied",     EQUIP_FLAG_READY },
	{ "Torso",       EQUIP_FLAG_TORSO },
	{ "Waist",       EQUIP_FLAG_WAIST },
	{ "Weapon Hand", EQUIP_FLAG_WEAPON },
	{ "WeaponHand",  EQUIP_FLAG_WEAPON },
	{ "Two handed",  EQUIP_FLAG_WEAPON },
	{ "Wrist",       EQUIP_FLAG_WRIST },
	{ NULL, 0 }
};

gint item_light_sources = 0;
GSList *visible_items = NULL;
static GHashTable *item_db;

static void item_db_free (void);
static void item_record_deallocate (gpointer key, gpointer value,
    gpointer user_data);
static void item_db_parse_option (item_t *item, const gchar *str);
static void item_deallocate (item_t *item);


/* =========================================================================
 = ITEM_DB_INIT
 =
 = Initialize the item database
 ======================================================================== */

void item_db_init (void)
{
	mudpro_db.items.filename = g_strdup_printf (
		"%s%citems.db", character.data_path, G_DIR_SEPARATOR);

	item_db = NULL;
	item_db_load ();
}


/* =========================================================================
 = ITEM_DB_CLEANUP
 =
 = Cleanup after item database
 ======================================================================== */

void item_db_cleanup (void)
{
	item_db_free ();
	item_inventory_list_free ();
	item_visible_list_free ();

	g_free (mudpro_db.items.filename);
}


/* =========================================================================
 = ITEM_DB_LOAD
 =
 = (Re)load item database from file
 ======================================================================== */

void item_db_load (void)
{
	gchar buf[STD_STRBUF];
	gchar *offset, *tmp;
	item_t *item = NULL;
	FILE *fp;

	if ((fp = fopen (mudpro_db.items.filename, "r")) == NULL)
	{
		printt ("Unable to open item database");
		return;
	}

	g_get_current_time (&mudpro_db.items.access);

	if (item_db != NULL)
		item_db_free ();

	item_db = g_hash_table_new (g_str_hash, g_str_equal);

	g_hash_table_freeze (item_db);

	while (fgets (buf, sizeof (buf), fp))
	{
		strchomp (buf);

		if (buf[0] == '#' || buf[0] == '\0')
		{
			item = NULL;
			continue;
		}

		if (isspace (buf[0]))
		{
			if (item != NULL)
				item_db_parse_option (item, buf);
			continue;
		}

		offset = buf;
		if ((tmp = get_token_as_str (&offset)) == NULL)
			continue;

		if (!strcasecmp (tmp, "Item"))
		{
			item = g_malloc0 (sizeof (item_t));

			if ((item->name = get_token_as_str (&offset)) == NULL)
			{
				g_free (item);
				item = NULL;
			}
			else
				g_hash_table_insert (item_db, item->name, item);
		}
		g_free (tmp);
	}

	g_hash_table_thaw (item_db);

	fclose (fp);
}


/* =========================================================================
 = ITEM_DB_SAVE
 =
 = Save item database to file
 ======================================================================== */

void item_db_save (void)
{
}


/* =========================================================================
 = ITEM_DB_LOOKUP
 =
 = Lookup item in database
 ======================================================================== */

item_t *item_db_lookup (gchar *str)
{
	return g_hash_table_lookup (item_db, str);
}


/* =========================================================================
 = ITEM_DB_FREE
 =
 = Free memory allocated to item database
 ======================================================================== */

static void item_db_free (void)
{
	g_hash_table_foreach (item_db, item_record_deallocate, NULL);
	g_hash_table_destroy (item_db);
	item_db = NULL;
}


/* =========================================================================
 = ITEM_RECORD_DEALLOCATE
 =
 = Free memory allocated to item record
 ======================================================================== */

static void item_record_deallocate (gpointer key, gpointer value,
	gpointer user_data)
{
	item_deallocate (value);
}


/* =========================================================================
 = ITEM_DB_PARSE_OPTION
 =
 = Parses item option and adds data to db record
 ======================================================================== */

static void item_db_parse_option (item_t *item, const gchar *str)
{
	gchar *offset, *option, *arg;
	glong value;

	if (!item)
		return;

	offset = option = arg = NULL;

	offset = (gchar *) str;
	if ((option = get_token_as_str (&offset)) == NULL)
		return;

	if (!strcasecmp (option, "Equip"))
	{
		key_value_t *e;

		g_free (option);

		if ((arg = get_token_as_str (&offset)) == NULL)
			return;

		for (e = equip_flags; e->key; e++)
		{
			if (strcasecmp (arg, e->key))
				continue;

			FlagON (item->equip, e->value);
			g_free (arg);
			return;
		}
	}

	else if (!strcasecmp (option, "Limit"))
	{
		g_free (option);
		value = get_token_as_long (&offset);
		item->limit = MAX (0, value);
	}

	else if (!strcasecmp (option, "Reserve"))
	{
		g_free (option);
		value = get_token_as_long (&offset);
		item->reserve = MAX (0, value);
	}

	else if (!strcasecmp (option, "SetFlag"))
	{
		key_value_t *i;

		g_free (option);

		if ((arg = get_token_as_str (&offset)) == NULL)
			return;

		for (i = item_flags; i->key; i++)
		{
			if (strcasecmp (arg, i->key))
				continue;

			FlagON (item->flags, i->value);
			g_free (arg);
			return;
		}
	}

	else if (!strcasecmp (option, "Value"))
	{
		g_free (option);
		value = get_token_as_long (&offset);
		item->value = MAX (0, value);
	}

	g_free (arg);
}


/* =========================================================================
 = ITEM_LIST_ADD
 =
 = Add item to the list
 ======================================================================== */

GSList *item_list_add (GSList *slist, const gchar *str)
{
	item_t *item, *control;
	gchar *pos, *pos2, *buf;
	gulong equip;
	gint quantity;

	g_assert (str != NULL);

	quantity = 1;
	equip    = 0;

	/* create local copy we can modify */
	buf = g_strdup (str);

	/* check if item is equipped */
	if ((pos = item_equip_parse ((gchar *) buf, &equip)) != NULL)
		*(pos-2) = '\0'; /* remove tag from string */

	pos = buf;

	if (isdigit (pos[0]))
	{
		quantity = atoi (pos);
		while (*pos && !isalpha (*pos)) pos++;

		/* remove plurality */
		pos2 = pos + strlen (pos)-1;
		if (*pos2 == 's') *pos2 = '\0';
	}

	/* get control record */
	control = item_db_lookup (pos);

	/* item already in list, update existing record */
	if ((item = item_list_lookup (slist, pos)) != NULL)
	{
		if (control)
		{
			item->flags = control->flags;
			item->equip = control->equip;
		}
		else
		{
			item->flags = 0;
			item->equip = equip;
		}
		item->quantity += quantity;
		item->armed = CLAMP (item->armed | equip, 0, 1);
		item->hidden = character.flag.searching;
		g_free (buf);
		return slist;
	}

	/* item not in list, allocate space for new one */
	item = g_malloc0 (sizeof (item_t));
	if (control)
	{
		item->flags = control->flags;
		item->equip = control->equip;
	}
	else
	{
		item->flags = 0;
		item->equip = equip;
	}
	item->name = g_strdup (pos);
	item->quantity += quantity;
	item->armed = CLAMP (item->armed | equip, 0, 1);
	item->hidden = character.flag.searching;
	g_free (buf);
	return g_slist_prepend (slist, item);
}


/* =========================================================================
 = ITEM_LIST_MULTIPLE_ADD
 =
 = Add multiple items to list, via comma separated list
 ======================================================================== */

GSList *item_list_multiple_add (GSList *slist, const gchar *item_list)
{
	gchar *offset, *item;

	offset = (gchar *) item_list;

	while ((item = get_token_as_str (&offset)) != NULL)
	{
		slist = item_list_add (slist, item);
		g_free (item);
	}

	return slist;
}


/* =========================================================================
 = ITEM_LIST_DEL
 =
 = Remove item from list
 ======================================================================== */

GSList *item_list_del (GSList *slist, const gchar *str)
{
	gchar *pos, *pos2, *buf;
	item_t *item;
	gint quantity;

	quantity = 1;

	/* create local copy we can modify */
	buf = g_strdup (str);
	pos = buf;

	if (isdigit (pos[0]))
	{
		quantity = atoi (pos);
		while (*pos && !isalpha (*pos)) pos++;

		/* remove plurality from item name */
        pos2 = pos + strlen (pos)-1;
		if (*pos2 == 's') *pos2 = '\0';
	}

	if ((item = item_list_lookup (slist, pos)) == NULL)
	{
		g_free (buf);
		return slist; /* item not in list */
	}

	if (item->quantity <= quantity)
	{
		/* remove item from list */
		slist = g_slist_remove (slist, item);
		item_deallocate (item);
	}
	else /* update item record */
		item->quantity -= quantity;

	g_free (buf);
	return slist;
}


/* =========================================================================
 = ITEM_LIST_MULTIPLE_DEL
 =
 = Remove multiple items from list, via comma separated list
 ======================================================================== */

GSList *item_list_multiple_del (GSList *slist, const gchar *item_list)
{
	gchar *offset, *item;

	offset = (gchar *) item_list;

	while ((item = get_token_as_str (&offset)) != NULL)
	{
		slist = item_list_del (slist, item);
		g_free (item);
	}

	return slist;
}


/* =========================================================================
 = ITEM_LIST_LOOKUP
 =
 = Lookup item in list
 ======================================================================== */

item_t *item_list_lookup (GSList *slist, gchar *name)
{
	GSList *node;
	item_t *item;

	if (!slist || !name)
		return NULL;

	for (node = slist; node; node = node->next)
	{
		item = node->data;

		if (!strcasecmp (item->name, name))
			return item;
	}

	return NULL;
}


/* =========================================================================
 = ITEM_LIST_FREE
 =
 = Free memory allocated to item list
 ======================================================================== */

void item_list_free (GSList *slist)
{
	GSList *node;

	for (node = slist; node; node = node->next)
		item_deallocate (node->data);
}


/* =========================================================================
 = ITEM_INVENTORY_LIST_BUILD
 =
 = Parse string for items and add them to inventory
 ======================================================================== */

void item_inventory_list_build (gchar *str)
{
	gchar *str_items_carried = "You are carrying ";
	gchar *str_keys_carried  = "You have the following keys: ";
	gchar *str_keys_none     = "You have no keys.";
	GString *item_list;
	gchar *pos;

	g_assert (str != NULL);

	if (character.inventory)
		item_inventory_list_free ();

	item_light_sources = 0;

	item_list = g_string_new (str);

	if ((pos = strstr (item_list->str, str_items_carried)) != NULL)
		item_list = g_string_erase (item_list, pos-item_list->str,
			strlen (str_items_carried));

	if ((pos = strstr (item_list->str, str_keys_carried)) != NULL)
	{
		item_list = g_string_erase (item_list, pos-item_list->str,
			strlen (str_keys_carried));
		item_list = g_string_insert_c (item_list, pos-item_list->str, ',');
	}
	else if ((pos = strstr (item_list->str, str_keys_none)) != NULL)
		item_list = g_string_erase (item_list, pos-item_list->str,
			strlen (str_keys_none));

	character.inventory = item_list_multiple_add (
		character.inventory, item_list->str);
	g_string_free (item_list, TRUE);
	item_light_sources = item_inventory_get_quantity_by_flag (
		ITEM_FLAG_USABLE | ITEM_FLAG_LIGHT);
}


/* =========================================================================
 = ITEM_INVENTORY_LIST_FREE
 =
 = Free data allocated to inventory list
 ======================================================================== */

void item_inventory_list_free (void)
{
	item_list_free (character.inventory);
	character.inventory = NULL;
}


/* =========================================================================
 = ITEM_INVENTORY_MANAGE
 =
 = Manage items in character inventory
 ======================================================================== */

void item_inventory_manage (void)
{
	GSList *node;
	item_t *control, *item;
	gint excess;

	for (node = character.inventory; node; node = node->next)
	{
		item = node->data;

		if ((control = item_db_lookup (item->name)) == NULL)
			continue; /* item not in control list */

		if (control->flags & ITEM_FLAG_AUTO_DROP)
		{
			if (control->flags & ITEM_FLAG_CURRENCY)
				command_send_va (CMD_DROP, "%d %s",
					item->quantity, item->name);
			else
				command_send_va (CMD_DROP, "%s", item->name);

			return;
		}

		if (control->limit) /* drop excess items */
		{
			/* take both limit and reserve into account */
			excess = MAX (0,
				item->quantity - MAX (control->limit, control->reserve));

			if (excess)
			{
				if (control->flags & ITEM_FLAG_CURRENCY)
					command_send_va (CMD_DROP, "%d %s", excess, item->name);
				else
					command_send_va (CMD_DROP, "%s", item->name);

				return;
			}
		}

		if (automap.location && !automap.lost && character.option.stash &&
			(automap.location->flags & ROOM_FLAG_STASH) &&
			(control->flags & ITEM_FLAG_STASH))
		{
			/* stash non-reserve items */

			excess = MAX (0, item->quantity - control->reserve);

			if (excess)
			{
				if (control->flags & ITEM_FLAG_CURRENCY)
					command_send_va (CMD_HIDE, "%d %s", excess, item->name);
				else
					command_send_va (CMD_HIDE, "%s", item->name);

				return;
			}
		}

		if ((control->flags & ITEM_FLAG_AUTO_ARM) && !item->armed)
		{
			if (combat.strategy.current && combat.strategy.current->weapon)
				continue;

			command_send_va (CMD_ARM, "%s", item->name);
			return;
		}

		if (!item->armed && navigation.room_dark &&
			(control->flags & (ITEM_FLAG_USABLE | ITEM_FLAG_LIGHT)))
		{
			/* determine if we should use light source or conserve it */
			if (character.option.reserve_light && !automap.lost &&
				character.light_mode > LIGHT_RESERVE_LOW)
				continue;

			/* disregard light mode when light sources is low */
			if (character.option.reserve_light && !automap.lost &&
				item_light_sources < 2)
				continue;

			command_send_va (CMD_USE, "%s", item->name);
			return;
		}

		if (item->armed && !navigation.room_dark &&
			(control->flags & (ITEM_FLAG_USABLE | ITEM_FLAG_LIGHT)))
		{
			/* determine if we should remove light source */

			if (!character.option.reserve_light || automap.lost ||
				command_pending (CMD_SEARCH))
				continue;

			if (character.light_mode == LIGHT_RESERVE_LOW &&
				item_light_sources > 1)
				continue;

			if (character.light_mode == LIGHT_RESERVE_MED
				&& character.state == STATE_NONE)
				continue;

			command_send_va (CMD_REMOVE, "%s", item->name);
			return;
		}
	}
}


/* =========================================================================
 = ITEM_INVENTORY_GET_QUANTITY_BY_FLAG
 =
 = Find quantity of item(s) by flag(s)
 ======================================================================== */

gint item_inventory_get_quantity_by_flag (gulong flags)
{
	GSList *node;
	item_t *item;
	gint quantity = 0;

	for (node = character.inventory; node; node = node->next)
	{
		item = node->data;

		if (item->flags & flags)
			quantity += item->quantity;
	}
	return quantity;
}


/* =========================================================================
 = ITEM_INVENTORY_GET_CASH_AMOUNT
 =
 = Returns total amount of cash on hand (in copper)
 ======================================================================== */

gulong item_inventory_get_cash_amount (void)
{
	GSList *node;
	item_t *item, *control;
	gulong amount = 0;

	for (node = character.inventory; node; node = node->next)
	{
		item = node->data;

		if ((control = item_db_lookup (item->name)) == NULL)
			continue; /* not in control list */

		if (!item->quantity ||
			!(control->flags & ITEM_FLAG_CURRENCY))
			continue;

		if (!control->value)
		{
			printt ("Monetary value for '%s' has not been set!", control->name);
			continue;
		}

		amount += (control->value * item->quantity);
	}
	return amount;
}


/* =========================================================================
 = ITEM_VISIBLE_LIST_BUILD
 =
 = Build list of visible items
 ======================================================================== */

void item_visible_list_build (gchar *str)
{
	gchar *pos;

	g_assert (str != NULL);

	if (visible_items)
		item_visible_list_free ();

	str = g_strdup (str + strlen ("You notice "));
	if ((pos = strstr (str, " here.")) != NULL)
		*pos = '\0';

	visible_items = item_list_multiple_add (visible_items, str);
	g_free (str);
}


/* =========================================================================
 = ITEM_VISIBLE_LIST_FREE
 =
 = Free memory allocated to visible item list
 ======================================================================== */

void item_visible_list_free (void)
{
	item_list_free (visible_items);
	visible_items = NULL;
}


/* =========================================================================
 = ITEM_EQUIP_PARSE
 =
 = Determine whether item is equipped, return tag offset
 ======================================================================== */

gchar *item_equip_parse (gchar *str, gulong *equip_flag)
{
	key_value_t *e;
	gchar *pos;

	g_assert (str != NULL);

	for (e = equip_flags; e->key; e++)
	{
		if ((pos = strstr (str, e->key)) != NULL)
		{
			if (equip_flag) *equip_flag = e->value;
			return pos;
		}
	}
	return NULL;
}


/* =========================================================================
 = ITEM_PICKUP_NEXT
 =
 = Get next item for retrieval
 ======================================================================== */

item_t *item_pickup_next (gint *quantity)
{
	GSList *node;
	item_t *item, *control, *inventory;
	gint tmp, limit;

	g_assert (quantity != NULL);

	for (node = visible_items; node; node = node->next)
	{
		item = node->data;

		if ((control = item_db_lookup (item->name)) == NULL)
			continue; /* item not in control list */

		/* attempt to disqualify item */
		if ((control->flags & ITEM_FLAG_AUTO_DROP) ||
			!(control->flags & ITEM_FLAG_AUTO_GET))
			continue;

		/* item is already hidden at a stash point */
		if (item->hidden && character.option.stash &&
			automap.location &&	(automap.location->flags & ROOM_FLAG_STASH))
			continue; /* FIXME: this should take reserve into account */

		if ((control->flags & ITEM_FLAG_CURRENCY) && !character.option.get_cash)
			continue; /* do not pickup cash */

		if (!(control->flags & ITEM_FLAG_CURRENCY) && !character.option.get_items)
			continue; /* do not pickup items */

		if (!control->limit)
		{
			/* no limit set, take everything */
			*quantity = item->quantity;
			return item;
		}

		inventory = item_list_lookup (character.inventory, item->name);
		limit = MAX (control->limit, control->reserve);

		if (inventory) /* take difference between inventory and limit */
			tmp = CLAMP (item->quantity, 0, MAX (0, limit - inventory->quantity));
		else
			tmp = CLAMP (item->quantity, 0, limit);

		if (tmp) *quantity = tmp;
		else continue; /* no quantity to take */

		*quantity = tmp;
		return item;
	}

	return NULL;
}


/* =========================================================================
 = ITEM_DEALLOCATE
 =
 = Deallocate the item
 ======================================================================== */

static void item_deallocate (item_t *item)
{
	g_assert (item != NULL);
	g_free (item->name);
	g_free (item);
}


/* =========================================================================
 = ITEM_LIGHT_SOURCE_ACTIVATE
 =
 = Activate an light source item
 ======================================================================== */

gboolean item_light_source_activate (void)
{
	GSList *node;
	item_t *item, *control;

	for (node = character.inventory; node; node = node->next)
	{
		item = node->data;

		if ((control = item_db_lookup (item->name)) == NULL)
			continue; /* not in control list */

		if (!item->armed &&
			(control->flags & (ITEM_FLAG_USABLE | ITEM_FLAG_LIGHT)))
		{
			command_send_va (CMD_USE, "%s", item->name);
			return TRUE;
		}
	}
	return FALSE;
}


/* =========================================================================
 = ITEM_LIGHT_SOURCE_DEACTIVATE
 =
 = Deactivate current light source
 ======================================================================== */

gboolean item_light_source_deactivate (void)
{
	GSList *node;
	item_t *item, *control;

	for (node = character.inventory; node; node = node->next)
	{
		item = node->data;

		if ((control = item_db_lookup (item->name)) == NULL)
			continue; /* not in control list */

		if (item->armed &&
			(control->flags & (ITEM_FLAG_USABLE | ITEM_FLAG_LIGHT)))
		{
			command_send_va (CMD_REMOVE, "%s", item->name);
			return TRUE;
		}
	}
	return FALSE;
}
